test/host/wasi_nn/wasi_nn.cpp

```diff
@@ -0,0 +1,323 @@
+                               UINT32_C(1), UINT32_C(0), SetInputEntryPtr},
+                           Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(), static_cast<uint32_t>(ErrNo::Success));
+  }
+  StorePtr += (TensorDim.size() * 4 + TensorData.size());
+
+  // compute test
+  WasmEdge::Host::WasiNNCompute WasiNNCompute(Env);
+  WasmEdge::Host::WasiNNCompute EmptyWasiNNCompute(EmptyEnv);
+  // context id exceed
+  {
+    EXPECT_TRUE(WasiNNCompute.run(
+        &MemInst, std::initializer_list<WasmEdge::ValVariant>{UINT32_C(3)},
+        Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(),
+              static_cast<uint32_t>(ErrNo::InvalidArgument));
+  }
+  // empty ctx compute
+  {
+    EXPECT_TRUE(EmptyWasiNNCompute.run(
+        &MemInst, std::initializer_list<WasmEdge::ValVariant>{UINT32_C(0)},
+        Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(), static_cast<uint32_t>(ErrNo::Busy));
+  }
+  // successful run
+  {
+    EXPECT_TRUE(WasiNNCompute.run(
+        &MemInst, std::initializer_list<WasmEdge::ValVariant>{UINT32_C(1)},
+        Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(), static_cast<uint32_t>(ErrNo::Success));
+  }
+
+  WasmEdge::Host::WasiNNGetOuput WasiNNGetOuput(Env);
+  // WasmEdge::Host::WasiNNGetOuput EmptyWasiNNGetOuput(EmptyEnv);
+  // no compute ctx request
+  {
+    EXPECT_TRUE(WasiNNGetOuput.run(
+        &MemInst,
+        std::initializer_list<WasmEdge::ValVariant>{
+            UINT32_C(0), UINT32_C(0), StorePtr, 65532, BuilderPtr},
+        Errno));
+    // FIXME should allow a request for output with no computation before?
+    EXPECT_EQ(Errno[0].get<int32_t>(), static_cast<uint32_t>(ErrNo::Success));
+    EXPECT_EQ(*MemInst.getPointer<uint32_t *>(BuilderPtr), UINT32_C(4004));
+  }
+  // output index exceed
+  {
+    EXPECT_TRUE(WasiNNGetOuput.run(
+        &MemInst,
+        std::initializer_list<WasmEdge::ValVariant>{
+            UINT32_C(0), UINT32_C(10), StorePtr, 65532, BuilderPtr},
+        Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(),
+              static_cast<uint32_t>(ErrNo::InvalidArgument));
+  }
+  // successful run
+  {
+    EXPECT_TRUE(WasiNNGetOuput.run(
+        &MemInst,
+        std::initializer_list<WasmEdge::ValVariant>{
+            UINT32_C(1), UINT32_C(0), StorePtr, 65532, BuilderPtr},
+        Errno));
+    EXPECT_EQ(Errno[0].get<int32_t>(), static_cast<uint32_t>(ErrNo::Success));
+    EXPECT_EQ(*MemInst.getPointer<uint32_t *>(BuilderPtr), UINT32_C(4004));
+    std::vector<float> OutputClassification(
+        MemInst.getPointer<float *>(StorePtr, 1001) + 1,
+        MemInst.getPointer<float *>(StorePtr, 1001) + 1001);
+    std::vector<size_t> SortedIndex, CorrectClasses{963, 762, 909, 926, 567};
+    SortedIndex = classSort<float>(OutputClassification);
+    // the probability of class i is placed at buffer[i]
+    for (size_t I = 0; I < CorrectClasses.size(); I++) {
+      EXPECT_EQ(SortedIndex[I], CorrectClasses[I]);
+    }
+  }
```

Comment:
Please add these tests:
1. The output buffer out of bounds.
2. The byte written buffer out of bounds.